logParser.js

```diff
@@ -29,57 +32,38 @@ setInterval(() => {
 }, pullInterval)
 
 //remove any WTB sections from seller message
-function filterWTS(wordsArray) {
-    if (!wordsArray.includes('WTB')) {
-        //base case, no wtb in msg
-        return wordsArray
-    } else {
-        var auction = wordsArray;
-        let trim = [];
-        auction.forEach((word, index) => {
-            if (word === 'WTB') {
-                trim[0] = index;
-                trim[1] = auction.indexOf('WTS', index);
-                if (trim[1] === -1) {
-                    trim[1] = auction.length;
-                }
-                auction.splice(trim[0], trim[1]-trim[0]);
-                return;
-            }
-        })
-        return filterWTS(auction);
+function filterWTS(auction_contents) {
+    const filteredWTS = [...auction_contents.matchAll(WTS_REGEX)];
+    let auctionWTS = "";
+    for (let section in filteredWTS) {
+        auctionWTS += filteredWTS[section][1].trim() + " ";
     }
+    return auctionWTS.toUpperCase().trim();
 }
 
 function parseLog(text, logServer) {
-    //trim timestamp
-    var auction = text.toUpperCase().slice(text.indexOf(']') + 2, text.length);
-    //split words into array
-    var words = auction.split(' ');
     //test if is auction
-    if (words[1] === 'AUCTIONS,') {
+    const auction_text = text.match(AUC_REGEX);
+    if (auction_text) {
         client.streamAuction(text.replace(/[|]+/g, '|'), logServer);
-        //trim single quotes
-        words[2] = words[2].slice(1);
-        words = filterWTS(words);
-        auction = words.join(' ');
-        if (words.length > 2) {
+        const auction_user = auction_text[1];
+        const auction_contents = auction_text[2];
+        const auctionWTS = filterWTS(auction_contents);
+        if (auctionWTS) {
             itemList.forEach(({item_name, user_id, user_name, price, server}) => {
```

Comment:
Humorously, this is exactly the same problem I solved with Aho-Corasick for the wiki lookup parsing. This works fine with a small itemList (few users creating watches) and it's possible it will never be a scale problem for us, but I will look into switching it to do a linear-time search using the same algorithm. No reason to do `O(n)` matches when it could be `O(1)`.

Reply:
Ok, I looked through this logic again and A-C might not work, or at least not with the current way the watches are stored. Since there could be multiple watches with the same keyword, or parts of keywords, and A-C stops matching for a section once it finds a match, it would interfere with finding all possible matches for all users. There may not be a great way around this using A-C.

Fortunately, I think the scale should be relatively small. We may also want to implement a watch quota so users can only watch some (reasonable) amount of keywords, something like 30 to start. That'd ensure that someone couldn't macro watching literally one million things and DoS the bot.

Reply:
Great idea!